perf/collector_libpfm.go

```diff
@@ -154,25 +178,104 @@ func (c *collector) registerEvent(config *unix.PerfEventAttr, name string, cgrou
 	}
 	return nil
 }
 
-func (c *collector) addEventFile(name string, cpu int, perfFile *os.File) {
+func (c *collector) addEventFile(name string, pmuType uint32, cpu int, perfFile *os.File) {
 	_, ok := c.cpuFiles[name]
 	if !ok {
-		c.cpuFiles[name] = map[int]readerCloser{}
+		c.cpuFiles[name] = map[uint32]map[int]readerCloser{}
 	}
-	c.cpuFiles[name][cpu] = perfFile
+
+	_, ok = c.cpuFiles[name][pmuType]
+	if !ok {
+		c.cpuFiles[name][pmuType] = map[int]readerCloser{}
+	}
+
+	c.cpuFiles[name][pmuType][cpu] = perfFile
 }
 
 func (c *collector) setupNonGrouped(name string, cgroup int) error {
-	if !isLibpfmInitialized {
-		return fmt.Errorf("libpfm4 is not initialized, cannot proceed with setting perf events up")
+	perfEventAttr, err := getPerfEventAttr(name)
+	if err != nil {
+		return err
+	}
+
+	return c.registerEvent(perfEventAttr, name, cgroup)
+}
+
+func createPerfEventAttr(event CustomEvent) *unix.PerfEventAttr {
+	length := len(event.Config)
+
+	config := &unix.PerfEventAttr{
+		Type:   event.Type,
+		Config: event.Config[0],
+	}
+	if length >= 2 {
+		config.Ext1 = event.Config[1]
+	}
+	if length == 3 {
+		config.Ext2 = event.Config[2]
 	}
 
-	klog.V(5).Infof("Setting up non-grouped perf event %s", name)
+	setAttributes(config)
+	klog.V(5).Infof("perf_event_attr struct prepared: %#v", config)
+	return config
+}
+
+func (c *collector) setupRawNonGroupedUncoreMemoryController(event *CustomEvent, cgroup int) error {
+	klog.V(5).Infof("Setting up non-grouped raw perf uncore memory controller event %#v", event)
+
+	if event.Type == perfTypeHardware {
+		// Register event for all memory controllers.
+		for _, pmu := range imcPMUs {
+			newEvent := CustomEvent{
+				Type:   pmu.Type,
+				Config: event.Config,
+				Name:   event.Name,
+			}
+			config := createPerfEventAttr(newEvent)
+			err := c.registerUncoreMemoryControllerEvent(config, string(newEvent.Name), cgroup, pmu.Cpus)
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	} else {
+		// Register event only for the memory controller.
+		config := createPerfEventAttr(*event)
+		cpus, err := imcPMUs.getCpus(event.Type)
+		if err != nil {
+			return err
+		}
+		return c.registerUncoreMemoryControllerEvent(config, string(event.Name), cgroup, cpus)
+	}
+}
+
+func (c *collector) registerUncoreMemoryControllerEvent(config *unix.PerfEventAttr, name string, cgroup int, cpus []uint32) error {
+	for _, cpu := range cpus {
+		pid, groupFd, flags := cgroup, -1, unix.PERF_FLAG_FD_CLOEXEC|unix.PERF_FLAG_PID_CGROUP
+		fd, err := unix.PerfEventOpen(config, pid, int(cpu), groupFd, flags)
+		if err != nil {
+			return fmt.Errorf("setting up perf event %#v failed: %q", config, err)
+		}
+		perfFile := os.NewFile(uintptr(fd), name)
+		if perfFile == nil {
+			return fmt.Errorf("unable to create os.File from file descriptor %#v", fd)
+		}
+
+		c.addEventFile(name, config.Type, int(cpu), perfFile)
+	}
+
+	return nil
+}
+
+func getPerfEventAttr(name string) (*unix.PerfEventAttr, error) {
```

Comment:
You allocate memory in line 280, then return Go pointer allocated in line 284 but in line 281 you defer call to `free()`. `perfEventAttr` becomes a [dangling pointer](https://en.wikipedia.org/wiki/Dangling_pointer). `free()` must be called after `perfEventAttr` is no longer used.